qualtran/_infra/data_types.py

```diff
@@ -0,0 +1,147 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import abc
+from typing import TypeVar, Union
+
+import attrs
+import sympy
+
+
+class QDType:
+    @property
+    @abc.abstractmethod
+    def num_qubits(self):
+        """Number of qubits required to represent a single instance of this data type."""
+
+
+QDTypeT = TypeVar("QDTypeT", bound=QDType)
+
+
+@attrs.frozen
+class QBit(QDType):
+    """A single qubit. The smallest addressable unit of quantum data."""
+
+    @property
+    def num_qubits(self):
+        return 1
+
+
+@attrs.frozen
+class QAny(QDType):
+    """Opaque bag-of-qbits type. Should be used sparingly"""
+
+    bitsize: Union[int, sympy.Expr]
+
+    def __attrs_post_init__(self):
+        if self.bitsize == 1:
+            raise ValueError("bitsize must be greater than 1. Use QBit() instead.")
+
+    @property
+    def num_qubits(self):
+        return self.bitsize
+
+
+@attrs.frozen
+class QInt(QDType):
+    """Integer of a given width bitsize.
```

Comment:
signed? twos compliment? I think we need to be specific.

Reply:
This is in the doc but hasn't made it into code yet. I'm thinking it would just an attribute.

Reply:
how do you decide what's an attribute (complimentiness?) vs a new type (signedness)

Reply:
I was thinking that complimentiness is more of a hardware detail (sort of like endianness) but it's a fine line given that we're implementing algorithms at such a low level anyway and don't have ALUs. It's probably not a clear boundary. Adding another signed integer type would be fine too and maybe more explicit given that certain chemistry algorithms assume one complement quite explicitly. Maybe this is the correct thing to do since we're actually using the sign bit explicitly in places. WDYT?